### PR TITLE
Function tests cleanup (part-3)

### PR DESCRIPTION
**Pull request by kichalla**
- Deletes the ValidationWebSite
- Moves RemoteAttribute related files to BasicWebSite

@rynowak @dougbu

### CODE REVIEW

File: samples/MvcSandbox/wwwroot/_references.js [hunk 1]
```diff
@@ -1,3 +1,7 @@
 ﻿/// <autosync enabled="true" />
 /// <reference path="../gulpfile.js" />
 /// <reference path="js/site.js" />
+/// <reference path="lib/bootstrap/dist/js/bootstrap.js" />
+/// <reference path="lib/jquery/dist/jquery.js" />
+/// <reference path="lib/jquery-validation/dist/jquery.validate.js" />
+/// <reference path="lib/jquery-validation-unobtrusive/jquery.validate.unobtrusive.js" />
```
Comment (ryanbrandenburg) [inline samples/MvcSandbox/wwwroot/_references.js]: Looks like this is the only change to MvcSandbox, is it just a holdover or do I not understand what this file does?

File: test/WebSites/BasicWebSite/Areas/Aria/Controllers/RemoteAttribute_HomeController.cs [hunk 2]
```diff
@@ -1,41 +1,45 @@
...
-    [Route("[Area]/[Controller]/[Action]", Order = -2)]
     public class RemoteAttribute_HomeController : Controller
     {
-        private static Person _person;
+        private static RemoteAttributeUser _user;
 
         [HttpGet]
         public IActionResult Create()
         {
+            //System.Diagnostics.Debugger.Launch();
```
Comment (ryanbrandenburg) [inline test/WebSites/BasicWebSite/Areas/Aria/Controllers/RemoteAttribute_HomeController.cs]: Commented code.

File: samples/MvcSandbox/wwwroot/_references.js (hunk 1 above)
Comment (kichalla) [inline samples/MvcSandbox/wwwroot/_references.js]: These were the changes that I saw when I opened up the MVC solution in VS...do you not see something similar when you open it?

File: test/WebSites/BasicWebSite/Areas/Aria/Controllers/RemoteAttribute_HomeController.cs (hunk 2 above)
Comment (kichalla) [inline test/WebSites/BasicWebSite/Areas/Aria/Controllers/RemoteAttribute_HomeController.cs]: oops...thanks!

File: samples/MvcSandbox/wwwroot/_references.js (hunk 1 above)
Comment (dougbu) [inline samples/MvcSandbox/wwwroot/_references.js]: This is fine. Let VS have its way.

**Comment by kichalla** [comment]
ping

**Comment by rynowak** [comment]
:shipit:

File: samples/UrlHelperSample/CustomUrlHelperFactory.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (dougbu) [inline samples/UrlHelperSample/CustomUrlHelperFactory.cs]: Bizarre. Looks like the dividing line between parts 2 and 3 is a bit fuzzy.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs [hunk 4]
```diff
@@ -23,12 +23,12 @@ public RemoteAttributeValidationTest(MvcTestFixture<ValidationWebSite.Startup> f
         public HttpClient Client { get; }
 
         [Theory]
-        [InlineData("Aria", "/Aria")]
+        [InlineData("Area1", "/Area1")]
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs]: Oh nos! What's wrong with an area named Aria?

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Create.cshtml [hunk 5]
```diff
@@ -1,10 +1,7 @@
-﻿@model ValidationWebSite.Models.Person
+﻿@model BasicWebSite.Models.RemoteAttributeUser
 @{
-    object areaObject;
-    ViewContext.ActionDescriptor.RouteValueDefaults.TryGetValue("area", out areaObject);
-    var areaName = (areaObject as string) ?? "root";
-
-    ViewBag.Title = "Create in " + areaName + " area.";
+    Layout = "/Areas/Area1/Views/RemoteAttribute_Home/_Layout.cshtml";
+    ViewBag.Title = "Create in Area1 area.";
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Create.cshtml]: This code purposefully _calculated_ the ambient area name. Leave it alone.

Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Create.cshtml]: Separately just use a relative path for the `Layout` -- `Layout = "_Layout.cshtml";`.

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml [hunk 6]
```diff
@@ -1,10 +1,7 @@
-﻿@model ValidationWebSite.Models.Person
+﻿@model BasicWebSite.Models.RemoteAttributeUser
 @{
-    object areaObject;
-    ViewContext.ActionDescriptor.RouteValueDefaults.TryGetValue("area", out areaObject);
-    var areaName = (areaObject as string) ?? "root";
-
-    ViewBag.Title = "Details in " + areaName + " area.";
+    Layout = "/Areas/Area1/Views/RemoteAttribute_Home/_Layout.cshtml";
+    ViewBag.Title = "Details in Area1 area.";
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: Again, don't remove the area name calculation and use a relative path for the `Layout`.

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml [hunk 7]
```diff
@@ -64,13 +61,5 @@
...
-    @Html.ActionLink("Create a new one", "Create") |
-    @if (string.Equals("Aria", areaName, StringComparison.OrdinalIgnoreCase))
-    {
-        @Html.RouteLink("Go back to home", "Home")
-    }
-    else
-    {
-        @Html.RouteLink("Go to Aria home", "AriaHome")
-    }
+    @Html.ActionLink("Create a new one", "Create") | @Html.RouteLink("Go to Area1 home", "AriaHome")
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: This view was previously used from another area. What happened to that?

Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: (One of the more interesting end-to-end complications with the `RemoteAttribute` relates to areas and specifically whether the original view and the verification action are or are not in the same area.)

File: test/WebSites/BasicWebSite/Controllers/RemoteAttribute_HomeController.cs [hunk 8]
```diff
@@ -1,16 +1,14 @@
...
 
+using BasicWebSite.Models;
 using Microsoft.AspNet.Mvc;
-using ValidationWebSite.Models;
 
-namespace ValidationWebSite.MyArea.Controllers
+namespace BasicWebSite.Controllers
 {
-    [Area("Aria")]
-    [Route("[Area]/[Controller]/[Action]", Order = -2)]
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Controllers/RemoteAttribute_HomeController.cs]: Why remove attribute routing here? This code illustrated a workaround for an interesting routing interaction.

File: test/WebSites/ValidationWebSite/bower.json [hunk 9]
```diff
@@ -1,21 +0,0 @@
-{
-    "name": "ValidationWebSite",
-    "description": "Web site demonstrating various validations.",
-    "private": true,
-    "dependencies": {
-        "jquery": "1.11.2",
-        "jquery-validation": "1.13.1",
-        "jquery-validation-unobtrusive": "3.2.2"
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/bower.json]: To confirm: Decision was to leave the `[RemoteAttribute]`s now in `BasicWebSite` unusable even when debugging the site or otherwise using a browser?

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs (hunk 4 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs]: The reason I did this was that we are moving files from different websites to BasicWebSite and thought that the structure of BasicWebSite should look in a manner which goes well with other tests targeting this website. Also why `Aria` anyway? :smile: The structure of areas was also inconsistent, like we named the value in the attribute as `Aria` but the folder as `MyArea`

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Create.cshtml (hunk 5 above)
Comment (kichalla) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Create.cshtml]: Previously these tests were using attribute routing. Was this on purpose? I ask this because the route template that they were using is already part of the defined conventional routes and so felt unnecessary to have them defined and also they interfere with other routes(top most order) causing other tests to fail. When I removed the attribute routes, this way of getting area detail does not work `ViewContext.ActionDescriptor.RouteValueDefaults.TryGetValue("area", out areaObject);`.
Also aren't we covering the link generation of attribute routes somewhere else and so we can avoid it here as its not a pivot in this case.

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml (hunk 7 above)
Comment (kichalla) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: Previously areas and non-area based controllers used the views in the root shared folder. But I felt that since these files are being moved to BasicWebSite, their structure doesn't make sense and also this is not a pivot here. This kind of scenario should have been covered in 'area' specific tests? 
Regarding RemoteAttribute and areas, the `DisplayFor` extensions on the individual properties of the model generate links and they should still behave as expected?

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Create.cshtml (hunk 5 above)
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Create.cshtml]: > Was this on purpose?

Yes. The aim for #3624 is test migration, not scenario removal or modification except when removing entire sites. We're spending a lot of time on this PR because you're making additional changes that aren't justified.

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml (hunk 7 above)
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: What links are you talking about w.r.t. `DisplayFor()`?

File: test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs (hunk 4 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs]: > The structure of areas was also inconsistent, like we named the value in the attribute as Aria but the folder as `MyArea`

Well, the attributes are far more important than the folders but agree consistency is good. Just thrown by changes that aren't cleanup or consistency improvements in this PR.

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml (hunk 7 above)
Comment (kichalla) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: For example:

```
[Remote(  
action: "IsIdAvailable",  
controller: "RemoteAttribute_Verify",  
areaName:"Area1",  
ErrorMessage = "/Area1/RemoteAttribute_Verify/IsIdAvailable rejects you.")]  
public string UserId3 { get; set; }  
```
and its corresponding
```
42     <dl class="dl-horizontal"> 
43         <dt> 
44             @Html.DisplayNameFor(model => model.UserId3) 
45         </dt> 
46 
 
47         <dd> 
48             @Html.DisplayFor(model => model.UserId3) 
49         </dd> 
50     </dl> 
```

File: test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs (hunk 4 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs]: The folder names were needed to be consistent in this case because the views now are in their own area and to help view location search. Previously the views were in the root shared folder.

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml (hunk 7 above)
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: `DisplayNameFor()` and `DisplayFor()` don't "generate links". I'm missing something but this is a side conversation. Please restore the existing code and scenarios, except where the migration breaks things (e.g. namespaces) or you're doing straightforward cleanup.

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml (hunk 6 above)
Comment (kichalla) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: :+1: about relative paths

File: test/WebSites/ValidationWebSite/bower.json (hunk 9 above)
Comment (kichalla) [inline test/WebSites/ValidationWebSite/bower.json]: Not sure what you mean...I added these files in basic website now and you should be able to launch it from browser and test it too.

**Comment by kichalla** [comment]
:up: :date:

File: test/WebSites/BasicWebSite/project.json [hunk 10]
```diff
@@ -15,5 +15,10 @@
         "dnx451": { },
         "dnxcore50": { }
     },
+    "exclude": [
+        "wwwroot",
+        "node_modules",
+        "bower_components"
+    ],
     "webroot": "wwwroot"
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/project.json]: Remove this line. No modern `project.json` file should contain a `webroot` setting.

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml (hunk 7 above)
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: This change and other changes to the scenarios should be reverted. I'm :ok: with improving the folder layout and fitting into the existing areas in BasicViews.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs (hunk 4 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/RemoteAttributeValidationTest.cs]: :ok:

File: test/Microsoft.AspNet.Mvc.IntegrationTests/TryValidateModelIntegrationTest.cs [hunk 11]
```diff
@@ -140,7 +140,7 @@ public void TryValidateModel_CollectionsModel_ReturnsErrorsForInvalidProperties(
         {
             var controller = new TestController();
             controller.ControllerContext = new ControllerContext(operationContext.ActionContext);
-            controller.ObjectValidator = ModelBindingTestHelper.GetObjectValidator();
+            controller.ObjectValidator = ModelBindingTestHelper.GetObjectValidator(operationContext.MetadataProvider);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/TryValidateModelIntegrationTest.cs]: Does this indicate you need to rebase or an unrelated change mixed into this PR?

File: test/WebSites/BasicWebSite/Areas/Area1/Controllers/RemoteAttribute_HomeController.cs [hunk 12]
```diff
@@ -19,23 +19,23 @@ public IActionResult Create()
...
 
-            _person = person;
+            _user = user;
             return RedirectToAction(nameof(Details));
         }
 
-        [Route("/[Area]", Name = "AriaHome", Order = -3)]
-        [Route("/[Area]/[Controller]/Index", Order = -2)]
+        [HttpGet]
+        [Route("/[area]/[controller]/[action]", Name = "Area1Home")]
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Controllers/RemoteAttribute_HomeController.cs]: Why does this action now have `[HttpGet]`? Why have the URLs which will reach this action changed?

File: test/WebSites/BasicWebSite/Views/RemoteAttribute_Home/Create.cshtml [hunk 13]
```diff
@@ -0,0 +1,75 @@
+﻿@model BasicWebSite.Models.RemoteAttributeUser
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Views/RemoteAttribute_Home/Create.cshtml]: Odd this and other views show up as additions. Is the old `ValidationWebSite` completely clean?

**Comment by dougbu** [comment]
:watch: for some missing / restored bits.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/TryValidateModelIntegrationTest.cs (hunk 11 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/TryValidateModelIntegrationTest.cs]: Actually this was a change that I had to make after rebasing. This change was triggered after the commit https://github.com/aspnet/Mvc/commit/96de1dbe4b914734cfe97f271230ff6c6ddcfa11

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml (hunk 7 above)
Comment (kichalla) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: This view is in its own area `Area1` as opposed to being in the root shared folder in which case this piece of code made sense but here its not required. Also this is not a pivot in this case right?

File: test/WebSites/BasicWebSite/Areas/Area1/Controllers/RemoteAttribute_HomeController.cs (hunk 12 above)
Comment (kichalla) [inline test/WebSites/BasicWebSite/Areas/Area1/Controllers/RemoteAttribute_HomeController.cs]: Ok will remove the `HttpGet` here(but it was however implied in earlier case) and regarding the URLs, earlier the action was named `Details` but the route `Index`, I changed it to keep it consistent...if there was an issue here the tests would have caught it so we are good here

File: test/Microsoft.AspNet.Mvc.IntegrationTests/TryValidateModelIntegrationTest.cs (hunk 11 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/TryValidateModelIntegrationTest.cs]: I see now: The comparison branch in this PR must not be up-to-date so this shows up as a change here.

File: test/WebSites/BasicWebSite/Controllers/RemoteAttribute_HomeController.cs [hunk 14]
```diff
@@ -19,23 +18,21 @@ public IActionResult Create()
...
+                return View(user);
             }
 
-            _person = person;
+            _user = user;
             return RedirectToAction(nameof(Details));
         }
 
-        [Route("/[Area]", Name = "AriaHome", Order = -3)]
-        [Route("/[Area]/[Controller]/Index", Order = -2)]
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Controllers/RemoteAttribute_HomeController.cs]: Odd Details view isn't failing. Should have a route named `AriaHome` or update the `RouteLink()` call to use name `Area1Home` and restore at least one `[Route]` here.

File: test/WebSites/BasicWebSite/Areas/Area1/Controllers/RemoteAttribute_HomeController.cs [hunk 15]
```diff
@@ -1,41 +1,40 @@
-// Copyright (c) .NET Foundation. All rights reserved.
+﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using BasicWebSite.Models;
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Controllers/RemoteAttribute_HomeController.cs]: Note GitHub is confused between the Area1 controller (previously the Aria controller) and the root controller. Comparing Area1 controller to old root controller.

File: test/WebSites/BasicWebSite/Controllers/RemoteAttribute_HomeController.cs (hunk 14 above)
Comment (dougbu) [inline test/WebSites/BasicWebSite/Controllers/RemoteAttribute_HomeController.cs]: Actually GitHub is comparing the root controller with old Aria controller. Missing named route is "Home" but that might not be used anymore? (Should be.)

File: test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml [hunk 16]
```diff
@@ -64,13 +64,5 @@
 </div>
 
 <p>
-    @Html.ActionLink("Create a new one", "Create") |
-    @if (string.Equals("Aria", areaName, StringComparison.OrdinalIgnoreCase))
-    {
-        @Html.RouteLink("Go back to home", "Home")
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: Since this is a view for `Area1` which used to be `Aria`, should restore fact this was linking to root action.

Comment (dougbu) [inline test/WebSites/BasicWebSite/Areas/Area1/Views/RemoteAttribute_Home/Details.cshtml]: FYI I think this was more about navigating around site in a browser than a test addition. As a sample, we want the site to be human-usable even more.

**Comment by kichalla** [comment]
:up: :date:

**Comment by dougbu** [comment]
:shipit:

**Comment by kichalla** [comment]
57b88baad0372d3fe3f4bc8b9534d2643ddbf2dd
